### PR TITLE
[programmable transactions] Add scaffold for `Upgrade` command

### PR DESCRIPTION
**Pull request by tzakian**
This adds the basic plumbing for the `Upgrade` command. It takes an (owned) `UpgradeTicket` as an argument and will produce a `ReceiptTicket` with the new package's ID along with the `UpgradeTicket`'s `UpgradeCap` ID.  

There's probably some places where I did stupid stuff since I'm still paging things in, so please let me know if I went off-track somewhere :)

### CODE REVIEW

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:433]: It looks like the `ExecutionContext` has a `StateView` which will allow us to read form the object store for this purpose. I believe this should give us what we want :)

Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: The state view is the `TemporaryStore`. I think you would need to somehow expose the object store directly? If I'm understanding correctly

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:467]: The config is in the `Context`, so it should be easy to check. We should add/have some error in `ExecutionErrorKind` for `NotYetSupported` or something similar

File: crates/sui-adapter/src/programmable_transactions/context.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/context.rs:372]: If you are going to do this, could you refactor `new_package` out? I'd rather not have two functions that push to `new_packages`.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:278]: nit, make `execute_move_upgrade` return a `Vec`, just to be a bit more inline/idiomatic

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:462]: I think we probably just want a general error for "this feature isn't supported". Might be a bad idea to start carving it out per feature.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:476]: You need to type check the `ticket_obj_val` before doing `bcs::from_bytes` or you are going to accidentally support type conversion for upgrade tickets... which would be bad :)

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:522]: This will read from the temporary store, but no other new packages should be in there yet. Is that okay?

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:581]: Is this more of a "signature" check? Not sure where this should live or not

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:596]: nit `invariant_violation!("A helpful message for your future self oncall")`

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:612]: ditto

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:613]: When does normalize error? this doesn't feel like the right error kind below

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:624]: What is the error here? Does it give more info? Should we pass that along to the user?
Maybe at least we could give it as a string for dev-inspect cases

Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:624]: `ExecutionError::new_with_source` is the function for doing so

File: crates/sui-types/src/move_package.rs
Comment (tnowacki) [inline crates/sui-types/src/move_package.rs:96]: what is this field?

Comment (tnowacki) [reply crates/sui-types/src/move_package.rs:96]: Maybe I'm just a fool and should look at the `sui::package` code lol

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:522]: Not sure, so I'll let @amnn weigh-in. 

Can you publish multiple packages in the same transaction? And even if so I'm guessing this would be OK since they should have different package IDs I'd suspect...

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:613]: Good spot! This only happens if the deserialization of the modules fail. I'll rework this and it looks like we can get rid of the `Result` all together.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: That's right, `TemporaryStore` is only going to include the input objects for the transaction + any objects that we dynamically load (could we use the `ObjectRuntime` to do this @tnowacki?) so you won't find the package to be upgraded in there by default.  In most cases, packages are not loaded this way, they are loaded when `MoveVM` asks to load a module (which goes via `SyncModuleCache`).

File: crates/sui-types/src/move_package.rs
Comment (amnn) [reply crates/sui-types/src/move_package.rs:96]: https://github.com/MystenLabs/sui/blob/73dcf13605dcbec42646b75451245794946da155/crates/sui-framework/sources/package.move#L57-L76

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:486]: We shouldn't do this until we have checked compatibility, otherwise we're going to pollute the loader unnecessarily.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:522]: You can indeed publish multiple packages in the same transaction, and that could cause an issue here, where someone publishes and then upgrades a package in one transaction.  It's a strange thing to do, but we should respond to it in a sensible way.  In this case, it should be possible to support by also checking the `context`'s list of `new_packages`, right (which may then need to become an `IndexMap`.

But @tnowacki, will the temporary store be able to read the original package being upgraded in the first place? It's not going to be a transaction input.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:598]: nit, weird indent.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (amnn) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:605]: nit: Is there a way for us to explain the nature of this error, like we do below for the individual compatibility issues?

File: crates/sui-protocol-config/src/lib.rs
Comment (amnn) [inline crates/sui-protocol-config/src/lib.rs:116]: tiny nit: Do we need the `_allowed`? `package_upgrades` is the feature.
```suggestion
    package_upgrades: bool,
```

**Review by amnn — COMMENTED** [review]
Looking great!  I would say, initially, prioritise getting the change to the `Command` interface out, along with the protocol config change and a test to show that if you try and call the upgrade command, it will fail.  We can follow up with the rest of the implementation and testing in another PR which won't need to go in before March 15th.

File: crates/sui-protocol-config/src/lib.rs
Comment (tzakian) [reply crates/sui-protocol-config/src/lib.rs:116]: I don't think we do. I'll rename it :)

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:605]: I'm pretty sure we can. `ExistingModuleNotInUpgradedPackage(Identifier)` or something like that with a corresponding explanation of "Existing module {name} in package not found in next version of package" (or something like that).

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:486]: Sounds good. I originally had it lower down, but lifetime issues made me want to move it up here. I'll fix it :)

**Comment by tzakian** [comment]
I'm going to split this PR into two: the scaffolding part (which we do need to get in ASAP) and the actual implementation. Once I do that I'll undraft this PR.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:486]: Actually one question I had here is do we want to run the normalizer before we perform verification? From what I could tell poking around the code it seemed like a "this is fine, but you should really run the bytecode verifier first before doing this" which would mean we'd want to verify and publish first?

**Comment by tzakian** [comment]
Split PR so this contains only the skeleton/scaffolding. The next PR will be working on the actual upgrade logic. 

Added a simple Rust test to make sure we fail with the correct error if we try and upgrade a package right now.

File: crates/sui-types/src/move_package.rs
Comment (amnn) [inline crates/sui-types/src/move_package.rs:35]: I think the changes to this file could technically go into your second PR, but not a big deal.

**Review by amnn — APPROVED** [review]
🚀  Yes!  This is great, I think this is good to go, awesome work @tzakian, thank you for the quick turnaround cc @rvantonder to crib off this PR for making the relevant changes to the `Publish` command. @awelc will take it from there to interpret the `deps` coming in during publish and conver them to a linkage table!

The aim is that by the upcoming cut, a package that is published has all the right things (linkage and type origin table) in place so that it can be upgraded once we land the support for that in future.

File: crates/sui-adapter/src/programmable_transactions/execution.rs
Comment (tzakian) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: Didn't realize cargo fmt doesn't format `let-else` yet!

**Pull request merged**
